### PR TITLE
Add date datetime flags to datetime

### PR DESCRIPTION
**Pull request by JoeZiminski**
Update @DATE, @DATETIME, @TIME tags so that there are key-value pairs in the name (e.g. date-20221222). Also add underscore before these tags if not provided by user.  The main new code / functionality additions are [here](https://github.com/neuroinformatics-unit/datashuttle/blob/0a4191c303ae58235f35f0826ec90810472dbde9/datashuttle/utils_mod/utils.py#L261) and tested [1](https://github.com/neuroinformatics-unit/datashuttle/blob/0a4191c303ae58235f35f0826ec90810472dbde9/tests/tests_integration/test_make_dirs.py#L253) [2](https://github.com/neuroinformatics-unit/datashuttle/blob/0a4191c303ae58235f35f0826ec90810472dbde9/tests/tests_unit/tests_unit.py#L17) 

#24

Some additional changes got joined also on this PR, including
- removing depreciated configs from CLI
- removing depreciated user-editable sub- / ses- prefixes
- change preview to dry run
- handle spaces in paths

### CODE REVIEW

File: datashuttle/datashuttle.py [hunk 1]
```diff
@@ -344,22 +326,12 @@ def make_config_file(
...
-                                    directory on this machine
-        :param use_ephys_behav_camera: create camera directory in ephys
-                                       behaviour
                                     directory on this machine
         :param use_behav:           create behav directory
-        :param use_behav_camera:    create camera directory in
-                                    behav directory
 
         NOTE: higher level directory settings will override lower level
               settings (e.g. if ephys_behav_camera=True
```
Comment (niksirbi) [inline datashuttle/datashuttle.py:337]: This example is no longer relevant

File: datashuttle/datashuttle.py [hunk 2]
```diff
@@ -344,22 +326,12 @@ def make_config_file(
...
-        :param use_ephys_behav:     create behav directory in ephys
-                                    directory on this machine
-        :param use_ephys_behav_camera: create camera directory in ephys
-                                       behaviour
                                     directory on this machine
         :param use_behav:           create behav directory
-        :param use_behav_camera:    create camera directory in
-                                    behav directory
 
         NOTE: higher level directory settings will override lower level
```
Comment (niksirbi) [inline datashuttle/datashuttle.py:336]: This note is no longer relevant

File: datashuttle/datashuttle.py [hunk 3]
```diff
@@ -127,44 +119,37 @@ def set_attributes_after_config_load(self):
...
     def make_sub_dir(
         self,
-        experiment_type: str,
         sub_names: Union[str, list],
         ses_names: Optional[Union[str, list]] = None,
-        dont_make_ses_tree: bool = False,
+        experiment_type: str = "all",
     ):
         """
         Make a subject directory in the data type directory. By default,
```
Comment (niksirbi) [inline datashuttle/datashuttle.py:127]: This is no longer accurate

File: datashuttle/datashuttle.py [hunk 4]
```diff
@@ -127,44 +119,37 @@ def set_attributes_after_config_load(self):
     @check_configs_set
     def make_sub_dir(
```
Comment (niksirbi) [inline datashuttle/datashuttle.py:120]: Add ability for datetime flags to this docstring

File: datashuttle/datashuttle.py [hunk 5]
```diff
@@ -617,36 +580,48 @@ def _make_directory_trees(
...
+            for ses in ses_names:
+
+                ses_path = self._join("local", [top_level_dir, sub, ses])
+
+                utils.make_dirs(ses_path)
+
+                self.make_experiment_type_folders(
+                    experiment_type, ses_path, "ses"
+                )
+
```
Comment (niksirbi) [inline datashuttle/datashuttle.py:602]: Consider printing a message at the end of this function. Sth like "Directories were successfully created". Currently it prints nothing if it's successfull. Optionally you could output all the paths that were created. This way the user will immediately see if they made a mistake

File: datashuttle/utils_mod/utils.py [hunk 6]
```diff
@@ -313,23 +258,55 @@ def process_names(
     return prefixed_names
 
 
-def update_ses_names_with_datetime(names: list):
+def update_names_with_datetime(names: list):
```
Comment (niksirbi) [inline datashuttle/utils_mod/utils.py:261]: A very easy mistake for the user to make is to type lowercase, e.g. `@date`. Consider either implicitly capitalising (being case-insensitive) or throwing a warning

File: datashuttle/datashuttle.py [hunk 7]
```diff
@@ -127,44 +119,37 @@ def set_attributes_after_config_load(self):
...
         :param ses_names:       session names (same format as subject name).
                                 If no session is provided, defaults to
                                 "ses-001".
-        :param dont_make_ses_tree:   option to make the entire session tree under
-                                the subject directory. If True, the subject
-                                directory only will be created.
+        :param experiment_type: The experiment_type to make the directory
+                                in (e.g. "ephys", "behav", "histology"). If
+                                "all" is selected, directory will be created
+                                for all data type.
```
Comment (niksirbi) [inline datashuttle/datashuttle.py:139]: for all experiment types enabled in config

File: datashuttle/datashuttle.py [hunk 8]
```diff
@@ -174,37 +159,37 @@ def make_sub_dir(
...
         sub_names: Union[str, list],
         ses_names: Union[str, list],
-        preview: bool = False,
+        experiment_type: str = "all",
+        dry_run: bool = False,
     ):
         """
         Upload data from a local machine to the remote project
         directory. In the case that a file / directory exists on
         the remote and local, the local will not be overwritten
```
Comment (niksirbi) [inline datashuttle/datashuttle.py:170]: maybe you wanted to say the remote file will not be overwritten

File: datashuttle/datashuttle.py [hunk 9]
```diff
@@ -174,37 +159,37 @@ def make_sub_dir(
...
         :param sub_names: a list of sub names as accepted in make_sub_dir().
                           "all" will search for all sub- directories in the
                           data type directory to upload.
         :param ses_names: a list of ses names as accepted in make_sub_dir().
                           "all" will search each sub- directory for
                           ses- directories and upload all.
-        :param preview: perform a dry-run of upload, to see which files
+        :param dry_run: perform a dry-run of upload, to see which files
                         are moved.
+        :param experiment_type: see make_sub_dir()
```
Comment (niksirbi) [inline datashuttle/datashuttle.py:181]: Much better for the user to be able to read the experiment_types here. You can still direct to make_sub_dir() for details

**Comment by niksirbi** [comment]
When uploading data from local to remote, `datashuttle` is happy to send empty folders, but when downloading, it only downloads folder if there is a file inside. Any particular reason for this asymmetry?

**Review by niksirbi — COMMENTED** [review]
Most of the comments I left have to do with better documentation. The new functionality seems to work as intended

File: datashuttle/datashuttle.py [hunk 10]
```diff
@@ -655,57 +630,128 @@ def _make_directory_trees(
     def _transfer_sub_ses_data(
```
Comment (niksirbi) [inline datashuttle/datashuttle.py:630]: At the end of this function, consider printing that a succesfull transfer has occuered.

File: datashuttle/utils_mod/utils.py [hunk 11]
```diff
@@ -313,23 +258,55 @@ def process_names(
...
-            names[i] = val.replace("@DATE", date)
+    date = str(datetime.datetime.now().date().strftime("%Y%m%d"))
+    format_date = f"date-{date}"
+
+    time_ = datetime.datetime.now().time().strftime("%H%M%S")
+    format_time = f"time-{time_}"
+
+    for i, name in enumerate(names):
+
+        if "@DATETIME" in name:  # must come first
```
Comment (adamltyson) [inline datashuttle/utils_mod/utils.py:275]: Is it worth storing these specific strings (e.g. `@DATETIME`) somewhere that can be imported, so we don't need to find & replace if we ever change them?

File: tests/tests_integration/test_filesystem_transfer.py [hunk 12]
```diff
@@ -47,7 +51,7 @@ def test_transfer_empty_folder_structure(
 
         test_utils.check_directory_tree_is_correct(
             project,
-            base_path_to_check,
+            os.path.join(base_path_to_check, project._top_level_dir_name),
```
Comment (adamltyson) [inline tests/tests_integration/test_filesystem_transfer.py:54]: Can this be pathlib-ed?

Comment (JoeZiminski) [reply tests/tests_integration/test_filesystem_transfer.py:54]: now complete in #56

File: datashuttle/datashuttle.py (hunk 10 above)
Comment (JoeZiminski) [reply datashuttle/datashuttle.py:630]: Nice idea, in the end this is handled by rclone that prints to console

File: datashuttle/datashuttle.py (hunk 5 above)
Comment (JoeZiminski) [reply datashuttle/datashuttle.py:602]: Nice idea, this is now logged and when the function is comleteled it says  ([commit](https://github.com/neuroinformatics-unit/datashuttle/pull/66/commits/a9f45f0f811729af675ee5e44f22fec189c4e9f4))

`         f"Finished making directories. For log of all created "
            f"directories, pleasee see {self._logging_path}"`

File: datashuttle/utils_mod/utils.py (hunk 11 above)
Comment (JoeZiminski) [reply datashuttle/utils_mod/utils.py:275]: Nice idea I think so, this is now implemented ([link to commit](https://github.com/neuroinformatics-unit/datashuttle/pull/66/commits/b1399dc78e3f4b21a4899ce2f74535a69301aa14))

**Comment by JoeZiminski** [comment]
Thanks both @niksirbi @adamltyson for your comments, these changes have now been implemented in #66

**Review by niksirbi — APPROVED** [review]

**Pull request merged**